File: bindings/javascript/packages/core/src/annealing.ts
/** Complete transverse-field Ising/QUBO annealing (Moonlab v1.2.1). */

import { getModule } from './wasm-loader';

export type AnnealSchedule = 'linear' | 'quadratic' | 'cosine';

export interface AnnealConfig {
  totalTime?: number;
  numSteps?: number;
  numSamples?: number;
  /** Zero asks Moonlab to assign and report a seed. */
  seed?: bigint;
  schedule?: AnnealSchedule;
  driverStrength?: number;
  problemStrength?: number;
  secondOrder?: boolean;
}

export interface AnnealResult {
  numQubits: number;
  effectiveSeed: bigint;
  bestBitstring: bigint;
  mostLikelyBitstring: bigint;
  groundBitstring: bigint;
  groundDegeneracy: number;
  bestEnergy: number;
  groundEnergy: number;
  expectedEnergy: number;
  successProbability: number;
  residualEnergy: number;
  problemGap: number;
  finalNorm: number;
  samples: bigint[];
  sampleEnergies: number[];
}

export interface AnnealingWasmModule {
  HEAPU8: Uint8Array;
  HEAPF64: Float64Array;
  _malloc(bytes: number): number;
  _free(ptr: number): void;
  _moonlab_anneal_ising_v1(
    n: number, h: number, j: number, offset: number,
    totalTime: number, steps: number, samples: number, seed: bigint,
    schedule: number, driver: number, problem: number, secondOrder: number,
    summary: number, samplesOut: number, energiesOut: number,
  ): number;
  _moonlab_anneal_qubo_v1(
    n: number, q: number, offset: number,
    totalTime: number, steps: number, samples: number, seed: bigint,
    schedule: number, driver: number, problem: number, secondOrder: number,
    summary: number, samplesOut: number, energiesOut: number,
  ): number;
}

const SUMMARY_BYTES = 96;
const scheduleCode: Record<AnnealSchedule, number> = {
  linear: 0, quadratic: 1, cosine: 2,
};

function resolved(config: AnnealConfig = {}) {
  const result = {
    totalTime: config.totalTime ?? 10,
    numSteps: config.numSteps ?? 1000,
    numSamples: config.numSamples ?? 1024,
    seed: config.seed ?? 0n,
    schedule: scheduleCode[config.schedule ?? 'cosine'],
    driverStrength: config.driverStrength ?? 1,
    problemStrength: config.problemStrength ?? 1,
    secondOrder: config.secondOrder ?? true,
  };
  if (!(result.totalTime > 0) || !Number.isInteger(result.numSteps) ||
      result.numSteps < 1 || result.numSteps > 10_000_000 ||
      !Number.isInteger(result.numSamples) || result.numSamples < 1 ||
      result.numSamples > 10_000_000 || !Number.isInteger(result.schedule) ||
      result.seed < 0n ||
      result.seed > 0xffff_ffff_ffff_ffffn ||
      !(result.driverStrength > 0) || !(result.problemStrength > 0)) {
    throw new RangeError('invalid annealing configuration');
  }
  return result;
}

function squareDimension(matrix: readonly number[]): number {
  const n = Math.sqrt(matrix.length);
  if (!Number.isInteger(n) || n < 1) {
    throw new RangeError('matrix must be non-empty and square');
  }
  return n;
}

function writeF64(mod: AnnealingWasmModule, values: readonly number[]): number {
  const ptr = mod._malloc(values.length * 8);
  if (!ptr) throw new Error('Moonlab WASM allocation failed');
  mod.HEAPF64.set(values, ptr >>> 3);
  return ptr;
}

function parseResult(mod: AnnealingWasmModule, summary: number,
                     samplesPtr: number, energiesPtr: number,
                     n: number, count: number): AnnealResult {
  const view = new DataView(mod.HEAPU8.buffer);
  const u64 = (offset: number) => view.getBigUint64(summary + offset, true);
  const f64 = (offset: number) => view.getFloat64(summary + offset, true);
  const samples: bigint[] = [];
  const sampleEnergies: number[] = [];
  for (let i = 0; i < count; i++) {
    samples.push(view.getBigUint64(samplesPtr + i * 8, true));
    sampleEnergies.push(view.getFloat64(energiesPtr + i * 8, true));
  }
  return {
    numQubits: n,
    effectiveSeed: u64(0),
    bestBitstring: u64(8),
    mostLikelyBitstring: u64(16),
    groundBitstring: u64(24),
    groundDegeneracy: view.getUint32(summary + 32, true),
    bestEnergy: f64(40),
    groundEnergy: f64(48),
    expectedEnergy: f64(56),
    successProbability: f64(64),
    residualEnergy: f64(72),
    problemGap: f64(80),
    finalNorm: f64(88),
    samples,
    sampleEnergies,
  };
}

function buffers(mod: AnnealingWasmModule, count: number) {
  const summary = mod._malloc(SUMMARY_BYTES);
  const samples = mod._malloc(count * 8);
  const energies = mod._malloc(count * 8);
  if (!summary || !samples || !energies) {
    if (summary) mod._free(summary);
    if (samples) mod._free(samples);
    if (energies) mod._free(energies);
    throw new Error('Moonlab WASM allocation failed');
  }
  return { summary, samples, energies };
}

export async function annealIsing(
  fields: readonly number[], couplings: readonly number[], offset = 0,
  config: AnnealConfig = {},
): Promise<AnnealResult> {
  return annealIsingWithModule(
    (await getModule()) as unknown as AnnealingWasmModule,
    fields, couplings, offset, config);
}

/** Synchronous dependency-injected variant for embedders and unit tests. */
export function annealIsingWithModule(mod: AnnealingWasmModule,
                                      fields: readonly number[],
                                      couplings: readonly number[],
                                      offset = 0,
                                      config: AnnealConfig = {}): AnnealResult {
  const n = squareDimension(couplings);
  if (fields.length !== n) throw new RangeError('fields length must match matrix');
  const c = resolved(config);
  const h = writeF64(mod, fields), j = writeF64(mod, couplings);
  const out = buffers(mod, c.numSamples);
  try {
    const rc = mod._moonlab_anneal_ising_v1(
      n, h, j, offset, c.totalTime, c.numSteps, c.numSamples, c.seed,
      c.schedule, c.driverStrength, c.problemStrength,
      c.secondOrder ? 1 : 0, out.summary, out.samples, out.energies);
    if (rc !== 0) throw new Error(`moonlab_anneal_ising_v1 failed: ${rc}`);
    return parseResult(mod, out.summary, out.samples, out.energies, n, c.numSamples);
  } finally {
    mod._free(h); mod._free(j);
    mod._free(out.summary); mod._free(out.samples); mod._free(out.energies);
  }
}

export async function annealQubo(
  qubo: readonly number[], offset = 0, config: AnnealConfig = {},
): Promise<AnnealResult> {
  return annealQuboWithModule(
    (await getModule()) as unknown as AnnealingWasmModule,
    qubo, offset, config);
}

/** Synchronous dependency-injected variant for embedders and unit tests. */
export function annealQuboWithModule(mod: AnnealingWasmModule,
                                     qubo: readonly number[],
                                     offset = 0,
                                     config: AnnealConfig = {}): AnnealResult {
  const n = squareDimension(qubo);
  const c = resolved(config);
  const q = writeF64(mod, qubo);
  const out = buffers(mod, c.numSamples);
  try {
    const rc = mod._moonlab_anneal_qubo_v1(
      n, q, offset, c.totalTime, c.numSteps, c.numSamples, c.seed,
      c.schedule, c.driverStrength, c.problemStrength,
      c.secondOrder ? 1 : 0, out.summary, out.samples, out.energies);
    if (rc !== 0) throw new Error(`moonlab_anneal_qubo_v1 failed: ${rc}`);
    return parseResult(mod, out.summary, out.samples, out.energies, n, c.numSamples);
  } finally {
    mod._free(q);
    mod._free(out.summary); mod._free(out.samples); mod._free(out.energies);
  }
}

File: bindings/rust/moonlab/src/annealing.rs
//! Complete transverse-field Ising/QUBO quantum annealing (Moonlab v1.2.1).

use crate::error::{QuantumError, Result};
use moonlab_sys::{
    moonlab_anneal_config_t, moonlab_anneal_result_best_bitstring,
    moonlab_anneal_result_best_energy, moonlab_anneal_result_effective_seed,
    moonlab_anneal_result_expected_energy, moonlab_anneal_result_final_norm,
    moonlab_anneal_result_free, moonlab_anneal_result_ground_bitstring,
    moonlab_anneal_result_ground_degeneracy, moonlab_anneal_result_ground_energy,
    moonlab_anneal_result_most_likely_bitstring, moonlab_anneal_result_num_qubits,
    moonlab_anneal_result_num_samples, moonlab_anneal_result_problem_gap,
    moonlab_anneal_result_residual_energy, moonlab_anneal_result_sample,
    moonlab_anneal_result_success_probability, moonlab_anneal_result_t,
    moonlab_quantum_anneal_ising, moonlab_quantum_anneal_qubo, moonlab_qubo_to_ising,
};
use std::ptr;

#[repr(i32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AnnealSchedule {
    Linear = 0,
    Quadratic = 1,
    Cosine = 2,
    Piecewise = 3,
}

#[derive(Debug, Clone)]
pub struct AnnealConfig {
    pub total_time: f64,
    pub num_steps: usize,
    pub num_samples: usize,
    pub seed: u64,
    pub schedule: AnnealSchedule,
    pub driver_strength: f64,
    pub problem_strength: f64,
    pub second_order: bool,
}

impl Default for AnnealConfig {
    fn default() -> Self {
        Self {
            total_time: 10.0,
            num_steps: 1000,
            num_samples: 1024,
            seed: 0,
            schedule: AnnealSchedule::Cosine,
            driver_strength: 1.0,
            problem_strength: 1.0,
            second_order: true,
        }
    }
}

impl AnnealConfig {
    fn ffi(&self) -> moonlab_anneal_config_t {
        moonlab_anneal_config_t {
            total_time: self.total_time,
            num_steps: self.num_steps,
            num_samples: self.num_samples,
            seed: self.seed,
            schedule: self.schedule as _,
            driver_strength: self.driver_strength,
            problem_strength: self.problem_strength,
            second_order: i32::from(self.second_order),
            schedule_points: ptr::null(),
            num_schedule_points: 0,
            reverse_anneal: 0,
            initial_bitstring: 0,
            reverse_s_target: 0.0,
            reverse_hold_fraction: 0.0,
            anneal_offsets: ptr::null(),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct AnnealResult {
    pub num_qubits: usize,
    pub effective_seed: u64,
    pub best_bitstring: u64,
    pub most_likely_bitstring: u64,
    pub ground_bitstring: u64,
    pub ground_degeneracy: usize,
    pub best_energy: f64,
    pub ground_energy: f64,
    pub expected_energy: f64,
    pub success_probability: f64,
    pub residual_energy: f64,
    pub problem_gap: f64,
    pub final_norm: f64,
    pub samples: Vec<u64>,
    pub sample_energies: Vec<f64>,
}

struct ResultGuard(*mut moonlab_anneal_result_t);
impl Drop for ResultGuard {
    fn drop(&mut self) {
        if !self.0.is_null() {
            unsafe { moonlab_anneal_result_free(self.0) };
            self.0 = ptr::null_mut();
        }
    }
}

fn collect(raw: *mut moonlab_anneal_result_t) -> Result<AnnealResult> {
    let guard = ResultGuard(raw);
    if guard.0.is_null() {
        return Err(QuantumError::Ffi("annealing returned NULL".to_string()));
    }
    let count = unsafe { moonlab_anneal_result_num_samples(guard.0) };
    let mut samples = Vec::with_capacity(count);
    let mut sample_energies = Vec::with_capacity(count);
    for index in 0..count {
        let mut bits = 0u64;
        let mut energy = 0.0;
        let rc = unsafe { moonlab_anneal_result_sample(guard.0, index, &mut bits, &mut energy) };
        if rc != 0 {
            return Err(QuantumError::Ffi(format!(
                "moonlab_anneal_result_sample({index}) rc={rc}"
            )));
        }
        samples.push(bits);
        sample_energies.push(energy);
    }
    Ok(unsafe {
        AnnealResult {
            num_qubits: moonlab_anneal_result_num_qubits(guard.0),
            effective_seed: moonlab_anneal_result_effective_seed(guard.0),
            best_bitstring: moonlab_anneal_result_best_bitstring(guard.0),
            most_likely_bitstring: moonlab_anneal_result_most_likely_bitstring(guard.0),
            ground_bitstring: moonlab_anneal_result_ground_bitstring(guard.0),
            ground_degeneracy: moonlab_anneal_result_ground_degeneracy(guard.0),
            best_energy: moonlab_anneal_result_best_energy(guard.0),
            ground_energy: moonlab_anneal_result_ground_energy(guard.0),
            expected_energy: moonlab_anneal_result_expected_energy(guard.0),
            success_probability: moonlab_anneal_result_success_probability(guard.0),
            residual_energy: moonlab_anneal_result_residual_energy(guard.0),
            problem_gap: moonlab_anneal_result_problem_gap(guard.0),
            final_norm: moonlab_anneal_result_final_norm(guard.0),
            samples,
            sample_energies,
        }
    })
}

fn square_dimension(matrix: &[f64]) -> Result<usize> {
    let n = (matrix.len() as f64).sqrt() as usize;
    if n == 0 || n.checked_mul(n) != Some(matrix.len()) {
        return Err(QuantumError::Ffi(
            "matrix must be non-empty and square".to_string(),
        ));
    }
    Ok(n)
}

pub fn anneal_ising(
    fields: &[f64],
    couplings: &[f64],
    offset: f64,
    config: &AnnealConfig,
) -> Result<AnnealResult> {
    let n = square_dimension(couplings)?;
    if fields.len() != n {
        return Err(QuantumError::Ffi(
            "fields length must match coupling matrix".to_string(),
        ));
    }
    let ffi_config = config.ffi();
    let mut raw = ptr::null_mut();
    let rc = unsafe {
        moonlab_quantum_anneal_ising(
            n,
            fields.as_ptr(),
            couplings.as_ptr(),
            offset,
            &ffi_config,
            &mut raw,
        )
    };
    if rc != 0 {
        return Err(QuantumError::Ffi(format!(
            "moonlab_quantum_anneal_ising rc={rc}"
        )));
    }
    collect(raw)
}

pub fn anneal_qubo(qubo: &[f64], offset: f64, config: &AnnealConfig) -> Result<AnnealResult> {
    let n = square_dimension(qubo)?;
    let ffi_config = config.ffi();
    let mut raw = ptr::null_mut();
    let rc =
        unsafe { moonlab_quantum_anneal_qubo(n, qubo.as_ptr(), offset, &ffi_config, &mut raw) };
    if rc != 0 {
        return Err(QuantumError::Ffi(format!(
            "moonlab_quantum_anneal_qubo rc={rc}"
        )));
    }
    collect(raw)
}

pub fn qubo_to_ising(qubo: &[f64], offset: f64) -> Result<(Vec<f64>, Vec<f64>, f64)> {
    let n = square_dimension(qubo)?;
    let mut h = vec![0.0; n];
    let mut j = vec![0.0; n * n];
    let mut converted_offset = 0.0;
    let rc = unsafe {
        moonlab_qubo_to_ising(
            n,
            qubo.as_ptr(),
            offset,
            h.as_mut_ptr(),
            j.as_mut_ptr(),
            &mut converted_offset,
        )
    };
    if rc != 0 {
        return Err(QuantumError::Ffi(format!("moonlab_qubo_to_ising rc={rc}")));
    }
    Ok((h, j, converted_offset))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn seeded_qubo_replays() {
        let config = AnnealConfig {
            total_time: 12.0,
            num_steps: 1200,
            num_samples: 128,
            seed: 0x1234_5678_9abc_def0,
            ..AnnealConfig::default()
        };
        let q = [-1.0, 1.0, 1.0, -1.0];
        let a = anneal_qubo(&q, 1.0, &config).unwrap();
        let b = anneal_qubo(&q, 1.0, &config).unwrap();
        assert_eq!(a.samples, b.samples);
        assert_eq!(a.sample_energies, b.sample_energies);
        assert_eq!(a.ground_degeneracy, 2);
        assert_eq!(a.ground_energy, 0.0);
        assert!(a.success_probability > 0.95);
    }

    #[test]
    fn qubo_conversion_is_exact() {
        let q = [-1.5, 0.7, 0.3, 2.0];
        let (h, j, off) = qubo_to_ising(&q, 0.25).unwrap();
        for bits in 0..4u64 {
            let x0 = (bits & 1) as f64;
            let x1 = ((bits >> 1) & 1) as f64;
            let q_energy = 0.25 + x0 * q[0] * x0 + x0 * q[1] * x1 + x1 * q[2] * x0 + x1 * q[3] * x1;
            let z0 = 1.0 - 2.0 * x0;
            let z1 = 1.0 - 2.0 * x1;
            let i_energy = off + h[0] * z0 + h[1] * z1 + j[1] * z0 * z1;
            assert!((q_energy - i_energy).abs() < 1e-13);
        }
    }
}
